pyproject.toml

```diff
@@ -276,3 +277,23 @@ doctest_optionflags = [
     "IGNORE_EXCEPTION_DETAIL",
     "ELLIPSIS",
 ]
+
+[tool.conda-lock]
+channels = [
+    "conda-forge",
+    "defaults",
+]
+platforms = [
+    "osx-arm64",
+    "linux-64",
+]
+channel-priority = "strict"
+name = "pudl-dev"
+
+[tool.conda-lock.dependencies]
+recordlinkage = {source = "pypi"}
+sphinx-autoapi = {source = "pypi"}
+pip = ">=23,<24"
+python = ">=3.11,<3.12"
+python-snappy = ">=0.6,<1"
+sqlite = ">=3.36,<4"
```

Comment:
This is all the configuration for `conda-lock`

With this `pyproject.toml` all of our dependencies except for `recordlinkage` and `sphinx-autoapi` are installed from `conda-forge` (and those two packages alone are singled out for installation from PyPI).

Our programmatic non-python dependencies (snappy, sqlite..) also get installed from `conda-forge`.

We could add `jupyterlab` and the other interactive dependencies from `devtools/environment.yml` into the `dev` optional dependencies above, and `pandoc` / `nodejs` here if we wanted, and eliminate `devtools/environment.yml` and probably our other `*-environment.yml` files.

Would need to add some infrastructure for updating the lockfile on a regular basis and get used to using it to install our environment, but it does seem to be working.

Reply:
Yes, that's right -- we need a version of the package that can only be sourced from PyPII.  I've requested to be made a maintainer on one, and someone from `conda-forge` has commented on the potential abandonment of `sphinx-autoapi` in the most recent bot PR, so hopefully they'll both be available from conda.  But at the same time it's good to know that in the event there's some package we can really only get from PyPI, there's a way to do it without breakage.

Reply:
And yes it would be really great to have `pyproject.toml` be the ONLY source file, with dependencies broken out into groups / extras that can be selectively installed from the lockfile.  I didn't figure it out but I suspect there's a way to categorize the conda-only dependencies so that they can also be associated with a particular grouping, so we could have `jupyterlab`, `pandoc, `nodejs` etc. be part of the `dev` or a `jupyter` group -- they'd be invisible to `pip` (which couldn't satisfy many of them) but available in the lockfile for installation with `mamba`